torch/autograd/gradcheck.py

```diff
@@ -201,6 +203,59 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
 
     return jacobian, reentrant, correct_grad_sizes, correct_grad_types
 
+def failed_batched_grad_test_msg(output_idx, input_idx, res, exp):
+    return f"""
+For output {output_idx} and input {input_idx}, batched grad computation failed.
+
+If you are adding a new operator, a fast workaround is to use gradcheck with
+`check_batched_grad=False` and filing an issue. If the test is autogenerated
+from e.g. common_method_invocations, there should be a deny list you can add
+the test to. If you're interested in making your new operator work with
+batched grad computation, read on.
+
+If you're modifying an existing operator that supports batched grad computation,
+or wish to make a new operator work with batched grad computation, please read
+the following.
+
+To compute batched grads (e.g., jacobians, hessians), we vmap over the backward
+computation. The most common failure case is if there is a 'vmap-incompatible
+operation' in the backward pass. Please see
+NOTE: [How to write vmap-compatible backward formulas]
+in the codebase for an explanation of how to fix this.
+
+Got:
+{res}
+
+Expected:
+{exp}
+""".strip()
+
+def test_batched_grad(fail_test, input, output, output_idx):
+    diff_input_list = list(iter_tensors(input, True))
+    grad = functools.partial(torch.autograd.grad, output, diff_input_list, retain_graph=True, allow_unused=True)
+
+    def vjp(v):
+        results = grad(v)
+        results = tuple(grad if grad is not None else
+                        torch.zeros([], dtype=output.dtype, device=output.device).expand(inp.shape)
+                        for grad, inp in zip(results, diff_input_list))
+        return results
+
+    grad_outputs = [torch.randn_like(output) for _ in range(2)]
+
+    expected = [vjp(gO) for gO in grad_outputs]
+    expected = [torch.stack(shards) for shards in zip(*expected)]
```

Comment:
This isn't really a gradcheck right? A gradcheck is defined by comparing the analytic and numeric result. What you're doing here is verifying the functoriality of vmap over vjap.

Reply:
(There's nothing wrong with this test, but not sure if it makes sense as an add on to gradcheck in this case.)

Reply:
That's correct. I can add a comment to this describing this.

One alternative to the approach here is to compute the full analytical jacobian via vmap. The problem I have with that is performance: it seems like it would do a bit more work than what is currently implemented in this PR. However, benchmarking shows that it's not that much more work: 191m (cpu time) vs 195m (cpu time, this PR) vs 198m (cpu time, if we computed the full jacobian)

Reply:
I do agree that this testing strategy is better. My point is mostly an organizational one.

Reply:
In an ideal world I would separate out the "batched grad" check from gradcheck and have separate tests for them. In PyTorch today this is a little difficult without writing a lot of code: we use gradcheck all over the place, but our only "structured" testing happens in test_nn (in the ModuleTests) and test_autograd (with common_method_invocations).
These two sources (test_nn and the common_method_invocations table, and soon to be OpInfo) are not comprehensive and we end up manually invoking `gradcheck` across our tests.

So an alternative could be to put the batched grad testing into its own helper function `check_vmap_over_vjp`, hunt down all of the callsites of gradcheck, and call `check_vmap_over_vjp` right after said gradchecks. There's around 430 places where this happens which will take me a while to hunt down, but on the upside it gives the right organizational structure.

Reply:
I think for now keeping the organization this way makes it so that it is easier to test things; I can see splitting the functionality up as a way to implement OpInfo based tests. I'll add a comment into the code saying something about this.